### ISSUE TITLE
overriding get_value does not appear to work

### ISSUE THREAD

**Issue by VVelox** [nrdvana/perl-Language-FormulaEngine#4]
So was just looking into this for writing something that can lookup values in a more complex manner and I can't seem to get it to work.

```
use Language::FormulaEngine;

my $vars= { foo => 1, bar => 3.14159265358979, baz => 42 };

package MyNamespace {

        use parent 'Language::FormulaEngine::Namespace::Default';

        sub get_value{
                return 3;
        }

        1;
};
my $engine= Language::FormulaEngine->new(namespace => MyNamespace->new);
my $formula= $engine->compile( 'foo' );
print $formula->($vars)."\n";
```

That should print "3", but instead prints "1" as it is never calling the get_value override.

**Comment by nrdvana** [comment]
It works if you add

    $engine->compiler->variables_via_namespace(1);

I guess it would make sense for that to be the default...  Originally I was aiming to make the default the fastest option, but then added the get_value for more flexibility.  I should at least document this more thoroughly.  It also leads to an inconsistency between ``->evaluate($expr, $vars)`` and ``->compile($expr)->($vars)``.

But then, I've also further complicated things by making ``variables_via_namespace`` change the compiled function to accept a Namespace as an argument rather than just the variables.

How about this for a solution:
  * Compile with ``->get_value`` by default unless get_value is the exact coderef of the default namespace in which case it optimizes to just use the vars directly.
  * Deprecate ``variables_via_namespace``
  * Add a new ``Compiler->output_api`` which could be set to
    * ``function_of_vars`` (the default, which merges arguments into the ``variables`` of a new temporary namespace, and calls get_value)
    *  or ``namespace_method`` which makes compiled expressions into methods that would run against a Namespace object

Would that be closer to what you'd expect from the module?

**Comment by nrdvana** [comment]
Does the current master meet your needs?  I'm preparing to release it.

**Comment by VVelox** [comment]
Sorry, been a bit distracted the last few days. I'll test it out tonight. :)

**Comment by VVelox** [comment]
So sat down to look at this. The default behavior in this case is not an issue. The issue was I was not seeing it in the docs under 'Language::FormulaEngine' or 'Language::FormulaEngine::Namespace#get_value' for it.

That said this mostly works.

```
use Language::FormulaEngine;
use Data::Dumper;

my $vars= { foo => 1, bar => 3.14159265358979, baz => 42 };

package MyNamespace {
        use parent 'Language::FormulaEngine::Namespace::Default';
        use Data::Dumper;
        sub get_value{
                print Dumper( \@_ );
                return 3;
        }

        1;
};
my $engine= Language::FormulaEngine->new(namespace => MyNamespace->new);
$engine->compiler->variables_via_namespace(1);
my $formula= $engine->compile( 'foo + A1' );

print $formula->($vars)."\n";
```

It prints 6 as expected, but $vars never gets properly passed.

```
$VAR1 = [
          bless( {
                   'variables' => {},
                   'constants' => {}
                 }, 'MyNamespace' ),
          'foo'
        ];
$VAR1 = [
          bless( {
                   'variables' => {},
                   'constants' => {}
                 }, 'MyNamespace' ),
          'A1'
        ];
6
```

So if you are attempting to pass a object such as a spread sheet to possibly pull variables from such as A1 or the like, you can't as $self->{variables} never gets populated.

**Comment by nrdvana** [comment]
If you enable ``variables_via_namespace`` it changes the compiled function to accept Namespace fields as the function arguments, so it would be

    print $formula->({ variables => $vars })."\n";

I realized how unintuitive this was as I was explaining it above, and that's why I decided to deprecate ``variables_via_namespace`` and replace it with ``output_api`` and ``optimize_var_access``.

If you try the current git master branch, you can remove this line

    $engine->compiler->variables_via_namespace(1);

and I think it should work for you.  If it does everything you need it to do, I'll release it to CPAN.

**Comment by VVelox** [comment]
Nice! Thanks! It works as expected.

```
$VAR1 = [
          bless( {
                   'constants' => {},
                   'variables' => {
                                    'baz' => 42,
                                    'bar' => '3.14159265358979',
                                    'foo' => 1
                                  }
                 }, 'MyNamespace' ),
          'foo'
        ];
$VAR1 = [
          bless( {
                   'constants' => {},
                   'variables' => {
                                    'baz' => 42,
                                    'bar' => '3.14159265358979',
                                    'foo' => 1
                                  }
                 }, 'MyNamespace' ),
          'A1'
        ];
6
```

**Comment by nrdvana** [comment]
Released as 0.06
